F-FrontEnd/src/F-datatype.c

```diff
@@ -1238,7 +1238,7 @@ type_is_compatible_for_allocation(TYPE_DESC left, TYPE_DESC right)
                               /*for_argument=*/FALSE,
                               /*for_assignment=*/TRUE,
                               /*is_pointer_assignment=*/TRUE,
-                              /*compare_rank=*/TRUE,
+                              /*compare_rank=*/FALSE,
```

Comment:
The source or mold must be conformable with the allocatable array. This conformability is not checked and no error is detected for the following code if the `compare_rank` flag is FALSE.

```fortran
program test
  real, allocatable :: a(:,:)
  real x(100)
  allocate (a(100,100), source=x)
END program test
```